Plate layout

will-moore opened the pull request:
This adds a Plate layout to the viewer, to support ome-zarr HCS data as proposed under https://github.com/ome/omero-ms-zarr/issues/73 and spec updated in https://github.com/ome/omero-ms-zarr/pull/75.

The spec is evolving and we are iteratively generating new sample datasets, using omero-cli-zarr.
I will try to keep this list of datasets up to date and working with this branch:

 - https://mystifying-lalande-e12142.netlify.app/?source=https://minio-dev.openmicroscopy.org/idr/idr0033-rohban-pathways/41744_illum_corrected/pr_43/5966.zarr (big plate, 9 fields per well)
 - https://mystifying-lalande-e12142.netlify.app/?source=https://minio-dev.openmicroscopy.org/idr/idr0002-heriche-condensation/plate1_1_013/pr_43/422.zarr (multi-T)
 - https://mystifying-lalande-e12142.netlify.app/?source=https://minio-dev.openmicroscopy.org/idr/idr0004-thorpe-rad52/P101/pr_43/1751.zarr (sparse plate)
 - https://mystifying-lalande-e12142.netlify.app/?source=https://minio-dev.openmicroscopy.org/idr/idr0001-graml-sysgro/JL_120731_S6A/pr_45/2551.zarr (multi-Z, multiple acquisitions)

To test:
 - Each of the above plates should show a plate layout (will show first field for each Well)
 - If multi-Z or multi-T, a slider should allow you to move the whole plate through Z/T
 - Clicking a Well should open a new browser tab showing the Well, showing all fields in the Well
 - If there are multiple acquisitions, you should be able to filter the Well fields by acquisition (adds `&acquisition=ID` to URL)
 - Clicking a field in the Well opens the Image itself in a new browser tab.

manzt commented:
Wow, that was fast! I'll hopefully be able to take a closer look later today, but just wanted to chime in to help clear out some of those console errors.

DeckGL layers use an `id` prop that must be unique to the layer. I think some of the errors right now could be coming from reusing the `id` for all of the wellplate layers. 

```javascript
const id = `${{layerProps.id}-plate-${x}-${y}`;
const wellProps = { id, ...layerProps, translate: [x, y], loader: layerProps.loaders[col + (row * layerProps.columns)]}
```

manzt commented:
@ilan-gold FYI

will-moore commented:
Thanks for the pointer @manzt. I'm not sure the `id` helped: I'm still seeing these warnings:
```
deck: Multiple new layers with same id ImageLayer({id: '72x12thwv8a'})
deck: Multiple old layers with same id ImageLayer({id: '72x12thwv8a'})
model.js?ca47:287 luma.gl: Found instanced attributes on non-instanced model XR-Static-Layer-0-64-84-0-0-72x12thwv8a
```
which are then leading to hundreds of identical errors:
```
deck: error during initialization of ImageLayer({id: '5q0qna2iqd2'}) Error: deck.gl: assertion failed.
```
Strangely, I don't see the custom `id` or even from the id `viv/ImageLayer` reflected in those errors.

All these errors are identical, until I pan or zoom the canvas, then I see them alternate with errors showing a different ID (from `viv/ImageLayer`):
```
deck: error during initialization of XRLayer({id: 'XR-Static-Layer-0-64-84-0-0-5q0qna2iqd2'}) Error: deck.gl: assertion failed.
```
All the above errors are coming from deck.gl layer.js
```
function _initState() {
      assert(!this.internalState && !this.state);
```

Also - I tried to add `onHover()` function to the layers but that seems to be failing (62dc631).
I would prefer to add `onClick()` but I don't see that supported by the `viv/ImageLayer`. Would be nice to have this so that clicking on a thumbnail can open a full viewer for that image.

Apologies for being a bit clueless here.
We're hoping to have a publicly available set of ome-zarr Plate data available shortly, so hopefully you could find time to try this yourself?

This is still very much exploring ways to show a Plate of images, and this is my first idea for how to do that. Any ideas for improving that would be much appreciated.
I'm currently testing with a locally-hosted 96-well plate (with down-sampled thumbnails for each well) and I'm feeling a bit of lag in zooming compared to a single image in vizarr. We want to be able to support 384-well plates (e.g. from idr0033) so I need to test with that soon to see how the performance is. It may be that loading separate thumbnails for each Well is just not going to scale well, and we might need to pre-generate a Plate view on the server to load as a single Image.
Many thanks!

ilan-gold commented:
@will-moore Can you make sure that the `ids` are unique?  Judging by these lines https://github.com/will-moore/vizarr/blob/55f3cc8afc9272a84bddf1732c0c1c0d0482c780/src/components/Viewer.tsx#L40-L49
it seems that the `range` iteration goes multiple times over the same `LayerProps` (so one `id` from the `LayerProps` for many iterations) whereas deck.gl needs unqiue ids.

As for `onClick` and `onHover`, is the argument in the `onHover` callback `undefined` or is it never being called?  We can add `onClick` as well, I think - I will add an issue for this.  I checked out the master branch of this codebase and added an `onHover` without any issue (`console.log` below), so it may have to do with the `id` issue - it's hard for me to say without an example:

<img width="1212" alt="Screen Shot 2020-10-21 at 9 38 58 AM" src="https://user-images.githubusercontent.com/43999641/96727669-530cf880-1381-11eb-9004-5194d77c9630.png">

ilan-gold commented:
Ah @will-moore you are overriding you `id` prop by spreading the `LayerProps` here: https://github.com/will-moore/vizarr/blob/55f3cc8afc9272a84bddf1732c0c1c0d0482c780/src/components/Viewer.tsx#L45.  This is likely this issue.

manzt commented:
@will-moore I'll be able to take a closer look on Friday. 

> Ah @will-moore you are overriding your id prop by spreading the LayerProps here: will-moore/vizarr@55f3cc8/src/components/Viewer.tsx#L45. This is likely this issue.

Yes, sorry for the bad bit of suggested code above. The ordering of keys with the `...` spread operator matters.

```javascript
// spread the parent object first and the add keys to override
const wellProps = { ...layerProps, id, translate: [x, y], loader: layerProps.loaders[col + (row * layerProps.columns)]}
```

will-moore commented:
Doh!
Yes, thanks both - the id fixed it, and the `onHover()` is working now too.
Adding `onClick()` support would be great, but I can work with `onHover()` for now.
 Thanks!

ilan-gold commented:
@will-moore `onClick` will be in the [next release](https://github.com/hms-dbmi/viv/pull/284) - thanks for the request.  Hoping to get that out ASAP.

will-moore commented:
Great, Thanks @ilan-gold!

will-moore commented:
I've uploaded my idr0002 test plate (96 wells, 2 channels) if you want to try:
http://localhost:3000/?source=https://minio-dev.openmicroscopy.org/idr/idr0002-heriche-condensation/plate1_1_013/422_no_T/422.zarr

The much bigger idr0033 plate (384 wells and 5 channels) is at https://minio-dev.openmicroscopy.org/idr/idr0033-rohban-pathways/41744_illum_corrected/5966_pr29.zarr but that is not displaying for me just now - will try to find where it's failing...

NB: These URLs are just for testing and could change at any time (as could the content/spec). Don't promote the links too widely etc. Thanks.

will-moore commented:
If I hard-code `rows = 8` in `io.js loadOMEPlate()` to reduce the size of the plate by 50%, then this does display - with an error:

![Screenshot 2020-10-22 at 11 43 16](https://user-images.githubusercontent.com/900055/96861473-0fd38800-145c-11eb-9b17-2d41e5bec961.png)

but without this, it just gets bogged down and never displays anything (fully black screen).

will-moore commented:
Sometimes it looks like the loading is held up by pending/failing `.zattrs` loading (get an error), but other times everything seems to load OK and still nothing is displayed.
Enabling only a single channel doesn't help (loading zarr pixel chunks is not the limiting factor).

will-moore commented:
One suggestion from @manics is that we could use https://zarr.readthedocs.io/en/stable/api/convenience.html#zarr.convenience.consolidate_metadata to reduce the number of calls for `.zarray` metadata?
But don't know if that's supported at-all by viv or vizarr (probably not?) or if the limit here is really the number of layers (384 too many)?

ilan-gold commented:
@will-moore I doubt this is a limitation of WebGL/deck.gl given my past experience (I have loaded 200MB datasets onto the GPU in WebGL, and 384 * 4 vertex attributes should not be an issue).  It's not crazy responsive, but I noticed that if you "refresh" the page (and your browser's cache is on) it works pretty reliably:
<img width="1679" alt="Screen Shot 2020-10-22 at 8 03 48 AM" src="https://user-images.githubusercontent.com/43999641/96869430-71d0c500-143d-11eb-96d4-abc1cb0d69ea.png">


Here it is zoomed in with updated rendering settings:
<img width="1675" alt="Screen Shot 2020-10-22 at 8 06 56 AM" src="https://user-images.githubusercontent.com/43999641/96869535-962ca180-143d-11eb-841d-7a8acd50734e.png">


I got a different errors, though, the first times I tried to load the dataset, one of which is more cryptic than the other:
<img width="1680" alt="Screen Shot 2020-10-22 at 8 00 56 AM" src="https://user-images.githubusercontent.com/43999641/96869599-ae9cbc00-143d-11eb-91a7-6784ed7ece07.png">
<img width="987" alt="Screen Shot 2020-10-22 at 8 17 20 AM" src="https://user-images.githubusercontent.com/43999641/96870621-2c14fc00-143f-11eb-8a97-eb0e3495f5f9.png">
The first looks like it is coming from the decoding step when `blosc` is running.  I think there may be something going on in the `zarr.js` internals we are not thinking of (since your error also came from there), but that is not really my wheelhouse, sadly.

As for consolidating metadata, I don't think I understand, but maybe @manzt does.  It seems like that is saying you should consider consolidating metadata on the backed, but I'm confused why we wouldn't be able to support it (although I don't know the `zarr.js` internals very well).  I guess it's that the `.zmetadata` file that link referenced is not supported yet in the browser `zarr.js` library?  Sorry I can't be of more help, but these issues do seem resolvable (given that it sometimes works) and I'm happy to continue to support when I can!

manzt commented:
I'll be able to take a closer look tomorrow.

The errors do look like something zarr.js specific, and on first glance likely not related to getting metadata. I've seen this cryptic error before and it's occurred when fetching chunks. I'm curious if it's somehow related to trying to make so many concurrent requests for data tiles.

ilan-gold commented:
@will-moore The new Viv package is up, version `0.5.0` with the `onClick` if you want to try upgrading.

will-moore commented:
@ilan-gold Thanks for the testing and the viv release. I'll give it a try...

will-moore commented:
I updated to `"@hms-dbmi/viv": "^0.5.0",` in my `package.json`, removed `/node_modules` and did `npm install`.
But I got:
```
Error: deck.gl - multiple versions detected: 8.2.8 vs 8.3.3
```
So I also updated package.json to `"deck.gl": "^8.3.3",` and re-clean + install.
That gave me:
```
layers-pass.js?1055:7 Uncaught Error: Model needs a program
    at assert (assert.js?9c7f:3)
    at Model._checkProgram (model.js?3e47:436)
    at Model.initialize (model.js?3e47:71)
```
So I deleted my `package-lock.json` and re-clean + install again...
This worked... Committed changes.

ilan-gold commented:
Sorry about not giving you a heads up on that @will-moore - I forgot this package includes its own `deck.gl` installation.

manzt commented:
Ok, so I did some investigating. Sorry about the delay. The cryptic error is from my mishandling of error propagation in the zarr.js source (apologies) https://github.com/gzuidhof/zarr.js/issues/64.

I should be able to get a patch out by the end of the day, but in the mean time if you find and replace:

`throw new error;` --> `throw error;` in `node_modules/zarr/dist/zarr.es6.js`

you will get more informative error messages. 

The cryptic error appears to be masking some network errors. I can't tell if it's from the client or the server throttling the number of requests. A known issue in `zarr.js` is that `openArray` will check if the store is a group (like zarr-python) making an unnecessary request for `.zgroup`. I could look into removing this to reduce the number of requests.

manzt commented:
@will-moore Not sure if you've been able to look into it more w.r.t. the zarr.js errors. Another option I've been thinking about here is to have a custom DeckGL layer for `vizarr` (e.g. `PlateLayer`/`HCSLayer`) that is just a deck.gl `CompositeLayer` like `ImageLayer` but specifically for a grid of images with the same rendering properties. That way we could push a lot of the grid logic outside `Viewer.ts` and into the layer itself. 

```javascript
layer = new PlateLayer({ id, loaders, loaderSelection, onClick etc... }); 
```

ilan-gold commented:
@manzt I would be fine with this especially since I am not sure we will actually get a way to scale/translate the `TileLayer` https://github.com/visgl/deck.gl/issues/5047.  We could then create a custom "hold-it-up-to-your-eye" layer like we have talked about.

I think we would probably want to bring the loader in as well, right @manzt?  I think we could bring in an OMETIFF version, but I would need an example dataset to work off of, if @will-moore had one: https://docs.openmicroscopy.org/ome-model/6.0.1/specifications/compliant-hcs.html

manzt commented:
@ilan-gold Vizarr isn't designed for OME-TIFF and won't be adding support since it is tied to OME-Zarr. This is just (potentially) a more performant way to add many `XRLayers` with the same render settings.

ilan-gold commented:
Ah @manzt I misread, my apologies.  I would still be in favor of having plate/HCS support in `viv` since it is/has been part of the `OME` data model.

joshmoore commented:
For what it's worth, there _will be_ a mapping from OME-Zarrs to OME-TIFFs (incl. for plates) if that means that this library should repository should only have the code for mapping into the model objects of viv.

manzt commented:
@joshmoore Good to know! My current concern is that we represent OME-TIFF and Zarr-based images in memory differently. The `loader` classes wrap different JavaScript objects, so we'd potentially need to think about a new loader for OME-TIFF plates.

We have been thinking about implementing the support for OME-TIFF as a custom Store (like how you would in python). This would unify the array-like objects we are dealing when loading data and rendering, and then other parts of viv could focus on mapping metadata.

will-moore commented:
@manzt Thanks for the error pointers. I tried those fixes locally. Mostly I just see some network requests timing out for the bigger plate. Maybe too many requests at once because if I subsequently visit the link of a failed request it works fine.
It eventually works with browser caching as @ilan-gold pointed out.

I like the idea of a Plate Layer (or a more general Grid Layer). I was wondering the best way of moving my Plate code out of Viewer component.
My ideal solution would be for a Multi-Resolution loader to know how to map tiles to OME-zarr Plate, so that when you zoom into a Plate grid, we can load higher resolution Well images for the Wells that are in the viewport (since we have them all pre-generated on the server).

For anyone who wants to try this PR without building locally, I pushed this branch to my master branch and am deploying that at https://mystifying-lalande-e12142.netlify.app/?source=https://minio-dev.openmicroscopy.org/idr/idr0002-heriche-condensation/plate1_1_013/422.zarr

NB: That plate is newly generated *with* Time data, so you can move the Plate through Time. Clicking on a Well opens full-resolution image in another Tab.

manzt commented:
The demo looks great. I hadn't even considered deep-linking, but that's exactly what the web is great for! I like the idea of being able to navigate an OME-Zarr hierarchy by linking to nested parts of the hierarchy, but agree that ideally a multi-resolution loader that is aware of the grid would be great.

The primary blocker is that we rely on deck.gl's `TileLayer` for multi-res which requires square tiles (`width === height`). This works for tiled pyramid (because the chunks for OME-Zarr/OME-TIFF are square), but since the lowest res of a grid image isn't necessarily square, I see issues with stitching together a multi-res grid. Certainly something I can experiment with.

ilan-gold commented:
@will-moore Would you mind if I shared your demo in support of a feature request for scale/translate for the base `TileLayer` in deck.gl?  I think it would help motivate the need/use-case more.

will-moore commented:
@manzt Yes, stitching Wells to create tiles might be tricky (I guess you could end up loading the same Well Image multiple times because parts of it are needed in several tiles), but if you wanted to enforce 1 tile == 1 Well (no stitching) then I think it would be OK to pad around a rectangular image in a square tile.

@ilan-gold I'm happy for you to share the links etc, but can't promise that anything will be working into the future. So maybe add a 'dev' warning etc. Hopefully things will be migrating to more permanent locations so you can update links in due course.

will-moore commented:
Another export from idr0033 plate:
https://mystifying-lalande-e12142.netlify.app/?source=https://minio-dev.openmicroscopy.org/idr/idr0033-rohban-pathways/41744_illum_corrected/pr_38/5966.zarr with rendering settings

will-moore commented:
Latest location for idr0033 plate: https://s3.embassy.ebi.ac.uk/idr/share/community-call-2020-10-29/idr0033-rohban-pathways/41744_illum_corrected/5966.zarr

oeway commented on src/components/Viewer.tsx:
```diff
@@ -27,11 +28,66 @@ function WrappedViewStateDeck({ layers }: { layers: Layer<VivLayerProps>[] }): J
+            translate: [x, y],
+            loader: layerInfo.loaders ? layerInfo.loaders[col + (row * columns)] : null,
+            pickable: true,
+            onClick: handleClick
```
It will very useful if we can override the default click handler for the ImJoy API, this will allow handle the click in other plugins in Python. Or use it as a plat explorer and when the user click on one we load the image into [ImageJ.JS](https://ij.imjoy.io).

Perhaps something like this:
 ```suggestion
            onClick: sourceInfo.click_callback || handleClick
```

manzt commented:
Hey will, sorry for the delay on this. I was working on a big build refactor for `viv` so that we have a pure ESM export. https://github.com/hms-dbmi/viv/pull/292

I forked this branch and have introduced `GridLayer`: https://github.com/hms-dbmi/vizarr/tree/grid-layer-plate-refactor. All the functionality is the same (I think), but now the `onClick` definition is isolated within `loadOMEPlate` and the `Viewer` has been refactored. Now each layer can be declared within `io.ts` and there is no special handling in `Viewer.ts`.

Since the click handler is now a part of `sourceData`, it should be straightforward to expose this to Imjoy @oeway. Thanks for the suggestion.

I have added a new dependency [`p-map`](https://github.com/sindresorhus/p-map), which you can set the number of concurrent requests on. Within grid layer, we can throttle how many requests go out each time. The only thing now is that tiles don't load visually until all data has been loaded; we can probably start to think about extending `TileLayer` rather than `CompositeLayer` but for now I think this is preferable.

@will-moore my preference would be that you checkout my branch and see what you think. Please make edits and merge here if you'd like.

will-moore commented:
With those last commits and build, previous links above won't be working since we are now expecting plate and well metadata as created by https://github.com/ome/omero-cli-zarr/pull/43
This link is now working and supports Plate -> Well -> Image browsing:
https://mystifying-lalande-e12142.netlify.app/?source=https://minio-dev.openmicroscopy.org/idr/idr0033-rohban-pathways/41744_illum_corrected/pr_43/5966.zarr

will-moore commented:
@manzt Thanks for that.
The zooming and panning for a big plate is definitely smoother with your build: https://5fa1c6d3b83eb4000765bc52--vizarr.netlify.app/?source=https://s3.embassy.ebi.ac.uk/idr/share/community-call-2020-10-29/idr0033-rohban-pathways/41744_illum_corrected/5966.zarr
And it's nice to have the click handlers created in the same place as the data. This also makes it possible for me to add different handlers for a Plate grid and for a grid of fields within a Well.
However, I recently noticed a warning for that plate `deck: Too many pickable layers, only picking the first 255` and it is still there with the GridLayer which means that the lower Wells aren't clickable.

A minor outstanding issue from my click handling is the encoding of row and column in the layer ID: The onClick functions need to know how the GridLayer encodes row & column in the layer ID, which feels like a brittle API, but it works for now.
It's a pity you don't see any tiles loading till they're all done, especially for the larger plate, but it's worth the wait.

@oeway I added an `onClick` to the imageConfig but I'm not sure how to create a plate from a notebook. Could you do that with a `source` URL for a plate?

At some point in the future we might want to be able to place "fields" at arbitrary coordinates within a Well. It looks like only a small part of the GridLayer logic is dealing with x and y coordinates, so I wonder if this could be moved inside the loadOMEPlate / loadOMEWell and we can pass the coordinates along with the loaders `loaders: [{x, y, w, h, loader}]`.
But we don't have any sample data ready for this yet.

manzt commented:
Agreed with brittle API and unfortunate with too many pickable layers. Perhaps there is a way to make just `GridLayer` pickable (rather than all `n` sublayers), and somehow hook inject `gridItem: { row: x, col: y}` into  the `info` object for the `onClick` handler.

> @oeway I added an onClick to the imageConfig but I'm not sure how to create a plate from a notebook. Could you do that with a source URL for a plate?

ilan-gold commented:
@will-moore Hopefully we can get around the picking problem when the `TileLayer` is part of the `GridLayer` as well - still waiting for a deck.gl release for that but the work should be done there.

manzt commented:
@will-moore, would try to work on this a bit more but struggling to focus due to the election.
In `gridLayer.ts`, you can edit `getPickingInfo` and perhaps compute the coordinate from there? Perhaps then we can turn off picking for the sublayers and just enable picking for `GridLayer`. This would negate the need to parse the layerId in the callback as well. You could just check `info?.gridCoord`.

```javascript
  getPickingInfo({ info, mode }: { info: any, mode: string }) {
    // other modes include "hover"; could inject this functionality for all actions
    if (mode === "query") {
      // triggered on click
      // `info` is what is provided in the `onClick` callback
      const { rows, columns, spacer } = this.props;
      const { width, height } = this.state;
      const row = // info.y - top ....
      const col = // ingo.x - left ...
      info.gridCoord = { row, col }; // add custom property
    }
    return info;
  }
```

will-moore commented:
Hi @mant,
 Thanks for the pointers. I updated `getPickingInfo()` to add `gridCoord` to the info. 0b57140dbce15e
But I couldn't see how to "turn off picking for the sublayers and just enable picking for `GridLayer`"?

We are starting to stabilise the HCS spec for ome-zarr. There's possibly some more changes to support acquisitions but this PR is looking like it will move from Draft to "Ready for review" at some point.

Are there any major issues you have with the way this PR is looking at the moment?
Do you have a test set of images / workflows to check for regression testing (I guess the example Jupyter notebooks are a good start, as well as the existing set of ome-zarr images). NB: we are planning on listing the current set of ome-zarr images (not HCS) in a blog post, with vizarr to show them: https://github.com/ome/blog/pull/67

Thanks!

manzt commented:
> But I couldn't see how to "turn off picking for the sublayers and just enable picking for `GridLayer`?

This is my mistake. I was under the impression that a `CompositeLayer` can be `pickable` independent of the sublayers. The suggested change in visgl/deck.gl#5113 is to make all sublayers (`XRLayer.pickable === false`) and then add a _single_ pickable layer behind the `n * m` `GridLayers`. The suggestion (in the issue) for this pickable layer is `GridCellLayer`, but since the grid shape is not square (`width !== height`), I don't think this will work. What we really need is a way to register click events (since we can compute `info.gridCoord` from `info.coordinate`).

My suggestion is to try the [`SolidPolygonLayer`](https://deck.gl/docs/api-reference/layers/solid-polygon-layer) and create a polygon (rectangle) that is the size of the entire grid. 

```javascript
renderLayers() {
    const { gridData, width, height } = this.state;
    if (width === 0 || height === 0) return null; // early return if no data

    const { loaders, rows, columns, spacer, id = '' } = this.props;
    const top = -(rows * (height + spacer)) / 2;
    const left = -(columns * (width + spacer)) / 2;
    const gridLayers = range(rows).flatMap((row) => {
      return range(columns).map((col) => {
        /* ... */
        const layerProps = {
          /* ... */
          pickable: false,
        };
        return new XRLayer({ ...this.props, ...layerProps });
      });
    });
    
    if (this.props.pickable) {
        const right = /* */;
        const bottom = /* */;
        const polygon = [[top, left], [top, right], [bottom, right], [bottom, left]];
        const layerProps = {
          data: [{ polygon }],
          getPolygon: d => d.polygon,
          getFillColor: [0, 0, 0, 0], // transparent
          getLineColor: [0, 0, 0, 0],
          pickable: true, // enable picking
        }
        const pickableLayer = new SolidPolygonLayer({ ...this.props, ...layerProps });
        return [...gridLayers, pickableLayer];
    }

    return gridLayers;
  }
```

> We are starting to stabilise the HCS spec for ome-zarr. There's possibly some more changes to support acquisitions but this PR is looking like it will move from Draft to "Ready for review" at some point.

Thanks for making me aware. Looking forward to reviewing. (PS. I noticed that the new `Acquisitions.tsx` didn't get pushed in the lastest, so I can't build the current branch locally).

> Are there any major issues you have with the way this PR is looking at the moment?

I don't see any major issues, other than that `vizarr` needs to make testing a priority with all the complexity that has amounted in `io.ts/ome.ts`. I guess the main issue is that we need stable fixtures `e.g. "a test set of images / workflows").

ilan-gold commented:
@manzt @will-moore Just published `0.7.0` of Viv which should allow you to layout the `MultiscaleImageLayer` on a grid so that you can zoom in on the grid and get the higher resolutions as you zoom in, like holding the plate up to you eye.  Thanks @manzt for bearing with me for the PR reviews :)

ilan-gold commented:
I should add this includes an upgrade of deck.gl so like last time, you'll need to upgrade the version hosted here.

will-moore commented:
Thanks @manzt - That pickable Polygon code worked great!

@ilan-gold Thanks for the `MultiscaleImageLayer` update. That will be really nice if we can get that working.

But I think we could maybe leave that for a future PR? This is already a big PR and I think it would be best to focus on getting this merged first? I'll ask others in the OME team to give this a good test and see if there's any functional fixes needed.

I recently updated a list of 4 Plates in the description above and added testing instructions. These should all be working now, and there's the list of images at https://blog.openmicroscopy.org/file-formats/community/2020/11/04/zarr-data/ that can also be used for testing. And also I need to test the example notebooks with this PR.

 Cheers

manzt commented:
> But I think we could maybe leave that for a future PR? This is already a big PR and I think it would be best to focus on getting this merged first?

Agreed. Lots of this PR deals with updating that state/data model for `vizarr` and I'd like to get that merged first. The multiscale changes will mostly be isolated to `gridLayer.ts`, and ideally a much smaller PR.

I'll hopefully have some time to give a thorough review this weekend. Thanks so much for pushing this forward, will

ilan-gold commented on src/gridLayer.ts:
```diff
@@ -0,0 +1,186 @@
+          id: `${id}-GridLayer-${row}-${col}`,
+          dtype: loaders[offset]?.dtype || '<u2', // fallback if missing,
+        };
+        return new XRLayer({ ...this.props, ...layerProps });
```
I know I'm a little late to the party but I was using the demo last night and noticed that it takes a while for the whole thing to load - was there a reason not to use `ImageLayer` so that each small image loads independently?  I might have missed something in the course of this PR that would prevent doing that.

ilan-gold replied:
Ok, maybe a future thing would be to wrap the loaders then so that the new `getTile` function calls the old one with a concurrency limit.  Future improvements :)

manzt commented on src/ome.ts:
```diff
@@ -0,0 +1,226 @@
+        let { source } = sourceData;
+        if (typeof source === 'string' && !isNaN(row) && !isNaN(column)) {
+            const field = (row * cols) + column;
+            if (source.endsWith('/')) {
+                source = source.slice(0, -1);
+            }
+            let imgSource = `${source}/${imagePaths[field]}`;
```
probably another place to use a util like `ensureEndsWith`:

```suggestion
            source = ensureEndsWith(source, '/');
            let imgSource = `${source}${imagePaths[field]}`;
```

will-moore replied:
I've created a util `rstrip(source, '/')` (same as python). I like seeing the `/` when creating paths like ```let imgSource = `${source}/${imagePaths[field]}`;``` since it helps to see the separate parts of the path.

manzt commented on src/ome.ts:
```diff
@@ -0,0 +1,226 @@
+    let columns: number = plateAttrs.columns.length;
+    let rowNames: string[] = plateAttrs.rows.map(row => row.name);
+    let columnNames: string[] = plateAttrs.columns.map(col => col.name);
+    let wellPaths = plateAttrs.wells.map(well => well.path + (well.path.endsWith('/') ? '' : '/'));
```
```suggestion
    let wellPaths = plateAttrs.wells.map(well => ensureEndsWith(well.path, '/'));
```

manzt commented on src/ome.ts:
```diff
@@ -0,0 +1,226 @@
+    async function getImageAttrs(path: string): Promise<any> {
+        if (path === '') return
+        try {
+            return await getJson(store, `${path}.zattrs`);
```
hmm, I guess the real util we want is some `join` method:

```javascript
await getJson(store, join(path, '.zattrs'));
```

That would remove the need to ensure individual strings have trailing (or leading) `/` and just ensure the correct path when making requests.

manzt replied:
What do you think? `ensureEndsWith` or `join`? Either way, having a consistent way of dealing with this would be ideal.

will-moore replied:
I've added a `join()` and `rstrip()` - belt and braces! But both are useful.

manzt commented on src/ome.ts:
```diff
@@ -0,0 +1,226 @@
+        } catch (err) {
+        }
+    }
+    for (let i = 0; i < imagePaths.length; i++) {
+        imageAttrs = await getImageAttrs(imagePaths[i]);
+        if (imageAttrs) {
+            break;
+        }
```
Are the `imageAttrs` the same, so we just need the first success?

will-moore replied:
The `imageAttrs` are assumed to be the same for each Well (except rendering settings) so we just need the first.

manzt commented on src/ome.ts:
```diff
@@ -0,0 +1,226 @@
+    let resolution = imageAttrs.multiscales[0].datasets.slice(-1)[0].path;
+
+    // Create loader for every Well. Some loaders may be undefined if Wells are missing.
+    async function createLoaderFromPath(path: string): Promise<ZarrLoader | undefined> {
+        try {
+            const data = await openArray({ store, path: `${path}${resolution}/` });
+            let l = createLoader([data]);
+            return l;
+        } catch (err) {
+            console.log(`Missing Well at ${path}`);
+        }
+    }
+    const promises = await pMap(imagePaths, createLoaderFromPath, { concurrency: 10 });
```
perhaps add concurrency as a default arg to `loadOMEPlate`? Also there is a top-level `createLoaderFromPath` we can use here I think:

```suggestion
    const mapper = path => createLoaderFromPath(store, join(path, resolution));
    const promises = await pMap(imagePaths, mapper, { concurrency });
```

will-moore replied:
Thanks - didn't think to use an extra `mapper` here.

manzt commented on src/ome.ts:
```diff
@@ -0,0 +1,226 @@
+    sourceData.rows = rows;
+    sourceData.columns = columns;
+    // Us onClick from image config or Open Well in new window
+    sourceData.onClick = config.onClick || ((info: any) => {
```
Should we also check for `config.onClick` like in `loadOMEWell`?

manzt commented on src/ome.ts:
```diff
@@ -0,0 +1,226 @@
+    sourceData.columns = columns;
+    // Us onClick from image config or Open Well in new window
+    sourceData.onClick = config.onClick || ((info: any) => {
+        let gridCoord = info.gridCoord;
+        if (!gridCoord) {
+            return;
+        }
+        const { row, column } = gridCoord;
+        let { source } = sourceData;
+        if (typeof source === 'string' && !isNaN(row) && !isNaN(column)) {
+            if (source.endsWith('/')) {
+                source = source.slice(0, -1);
+            }
+            let imgSource = `${source}/${acquisitionPath}${rowNames[row]}/${columnNames[column]}/`;
+            window.open(window.location.origin + '?source=' + imgSource);
+        }
+    })
+    return sourceData;
```
It seems like the difference between this callback and the one defined within `loadOMEWell` is just `imgSource`. Probably hard to make a shared function because both rely heavily on being closures, though.

manzt commented on src/state.ts:
```diff
@@ -38,8 +41,19 @@ export type SingleChannelConfig = {
+  rows?: number;
+  columns?: number;
+  acquisitions?: (string | Acquisition)[];
+  acquisition?: string,
```
Not `string | Acquisition`?

manzt commented on src/types/rootAttrs.d.ts:
```diff
@@ -20,6 +20,18 @@ export type OmeroImageData = {
   name?: string;
 };
 
+export type OmePlateData = {
+  name: string;
+  rows: { name: string }[];
+  columns: { name: string }[];
+  wells: { path: string } []
+  acquisitions?: { path: string }[]
+}
+
+export type OmeWellData = {
+  images?: { path: string, acquisition?: number }[]
+}
+
```
Great, thank you!

manzt reviewed (changes requested):
Made some comments mostly about creating some shared utilities and asking for some clarifications with regard to `Acquisitions`. Great idea to isolate the metadata parsing for `ome-zarr` with `ome.ts`.

I will checkout and test the Imjoy notebooks. Glad to have more public OME examples to test the new features here.

joshmoore commented:
> * (adds `&acquisition=ID` to URL)

by `index`? Also, what is intended to happen when `index > max_fields`, I'm currently getting a plate loaded anyway. (Though looking at it again, perhaps no `&acquisition` value gets me one plate, and a value gets me another, but otherwise it doesn't seem to be selecting.

joshmoore commented:
For https://mystifying-lalande-e12142.netlify.app/?source=https://minio-dev.openmicroscopy.org/idr/idr0004-thorpe-rad52/P101/pr_43/1751.zarr/0/E/5/, it looks like the Z slider is only modifying the YFP channel.

will-moore commented:
@joshmoore For https://mystifying-lalande-e12142.netlify.app/?source=https://minio-dev.openmicroscopy.org/idr/idr0004-thorpe-rad52/P101/pr_43/1751.zarr/0/E/5/ that looks like a bug in `omero-cli-zarr` export. The image only has 1 plane for that channel but it is exported for all z planes.

will-moore commented:
@joshmoore `&acquisition=` only filters the Well, not the Plate. See description "If there are multiple acquisitions, you should be able to filter the Well fields by acquisition (adds &acquisition=ID to URL)" When viewing a Plate, we don't know the fields in each Well, so we can't filter by acquisition, we can only show the first field (assume there is a field `0`).

joshmoore commented:
All understood, thanks, @will-moore. Are there any more commits coming here?

@manzt, what do you see from your side as the barriers to merging? With https://github.com/hms-dbmi/vizarr/pull/49, we will start shipping working versions along with the data we're generating. :tada: (Thanks)

manzt commented:
@joshmoore, just some minor comments that I'd like addressed but nothing major!

will-moore commented on src/components/LayerController/AcquisitionController.tsx:
```diff
@@ -0,0 +1,49 @@
+import { sourceInfoState } from '../../state';
+
+
+function AcquisitionController({ layerId }: {layerId: string,}): JSX.Element | null {
```
The `AcquisitionController` is only used when we have multiple `acquisitions` for a Plate. Each Well would have multiple fields and each acquisition (name and timestamp) might be linked to 1 or more field from each Well.
When viewing a Well with multiple acquisitions, this component allows you to filter by acquisition, e.g. https://mystifying-lalande-e12142.netlify.app/?source=https://minio-dev.openmicroscopy.org/idr/idr0001-graml-sysgro/JL_120731_S6A/pr_45/2551.zarr/B/1/. Filtering opens a window with same URL and `?acquisition=123`.
This is quite simplistic for now (doesn't update state and use router etc) but that could be a future improvement.

manzt replied:
Ok, I think I understand, thanks for the explanation! I ran into a _slight_ issue when performing the action:

- Select acquisition from drop down (Open drop-down and select `Acquisition: X`)
- Try to go back to original view (Open drop-down and select `Filter by Acquisition`)

This action adds `?aquistion=-1` to the URL, and I'm met with a blank screen. I'm ok with tracking this in an issue, but if you have an idea for how how to get the desired behavior I'd love to see that here!

will-moore commented on src/io.ts:
```diff
@@ -167,11 +136,26 @@ export async function createSourceData(config: ImageLayerConfig): Promise<Source
-      throw Error(`Failed to open arrays in zarr.Group. Make sure group implements multiscales extension.`);
+      // No rootAttrs in this group.
+      // if url is to a plate/acquisition/ check parent dir for 'plate' zattrs
+      const url: string = store.url.endsWith('/') ? store.url.slice(0, -1) : store.url;
```
The `store` here is what's returned from `const store = normalizeStore(source);` above, so it should be `HTTPStore`.

will-moore commented on src/io.ts:
```diff
@@ -167,11 +136,26 @@ export async function createSourceData(config: ImageLayerConfig): Promise<Source
+      const parentStore = normalizeStore(parentUrl);
+      const parentAttrs = (await getJson(parentStore, '.zattrs'));
+      if (parentAttrs?.plate) {
+        const acquisition = url.slice(url.lastIndexOf('/') + 1);
```
Sorry - this was left from an earlier version of ome-zarr HCS spec when acquisition was part of the plate/acquisition/row/column hierarchy and you could open `source=path/to/plate/acquisition` but not needed now - removing...

will-moore commented on src/io.ts:
```diff
@@ -235,13 +232,18 @@ export function initLayerStateFromSource(sourceData: SourceData, layerId: string
       id: layerId,
       loader,
+      loaders,
+      source,
```
`source` isn't needed but `loaders` used in `gridLayer` with:
```
const loaderChanged = typeof propsChanged === 'string' && propsChanged.includes('props.loaders');
```

will-moore commented:
@manzt I think I've addressed the issues above. Thanks for the feedback. Let me know if I've missed anything or any other fixes needed.

After merging origin/master above, the build started failing  (https://app.netlify.com/sites/mystifying-lalande-e12142/deploys/5fb539eaa3663900079ef41a) - my manual conflict resolution of `package.json` and `package-lock.json` is probably a bit dodgy, but even removing `package-lock.json` didn't help.
Something to do with https://www.npmjs.com/package/@danmarshall/deckgl-typings not being correct version (although their docs seem to indicate we're compatible)

So I pushed the pre-merge commit to https://github.com/will-moore/vizarr/commits/master to fix the build for now.

manzt reviewed (approved):
@will-moore thanks for taking the time to make these changes. I made one last comment about acquisitions and then I think I'll be able to merge.

I've had some issues with the `deck.gl` typings as well, but I just checked out this branch and built cleanly with `npm ci`. I'm not sure if I understand how you resolved the merge conflict, but if you're comfortable with merging I'm happy to for now and resolve this in a subsequent PR. I will make a PR to setup a new dev environment and hopefully netlify CD so that we can preview PRs here (just need to talk to Nils).

manzt commented:
Ok, I built in a clean env and tried out the imjoy examples. All work except the `mandelbrot` example. This is due to the viv upgrade in this PR where sometimes we call `z.store.getItem()` with more than one argument for [better performance with deck.gl 8.3.0](https://github.com/hms-dbmi/viv/pull/280) using `AbortSignal`s. A quick fix is to add an extra `options` argument to the zarr codec in `examples/imjoy_plugin.py` and set a default of `None`, 

```python
def encode_zarr_store(zobj):
    path_prefix = f"{zobj.path}/" if zobj.path else ""

    def getItem(key, options = None):
        return zobj.store[path_prefix + key]

    def setItem(key, value):
        zobj.store[path_prefix + key] = value

    def containsItem(key, options = None):
        if path_prefix + key in zobj.store:
            return True

    return {
        "_rintf": True,
        "_rtype": "zarr-array" if isinstance(zobj, zarr.Array) else "zarr-group",
        "getItem": getItem,
        "setItem": setItem,
        "containsItem": containsItem,
    }
```

@oeway any thoughts here? We can try to be more careful handling stores in `viv`, but in JavaScript providing more arguments to a function than a definition doesn't result in an error.

will-moore commented:
Thanks @manzt for catching that. Pushed fixes now to ignore invalid `&acquisition=ID` and also to not set `&acquisition=-1` when removing the filter.
Hope the build issues aren't too painful, but it would be great to have a PRs preview build.
Sorry it took a few days to respond - somehow I missed any notifications from this PR.
 Cheers!

manzt commented:
Great, thank you! So happy to have this PR in and thanks for working so hard to get it merged!

joshmoore commented:
@manzt : do you foresee this being part of an upcoming tag?

manzt commented:
@joshmoore -- done!

joshmoore commented:
Thanks, @manzt!
